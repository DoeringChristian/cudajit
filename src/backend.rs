pub trait Backend {
    type Compiler;
}
