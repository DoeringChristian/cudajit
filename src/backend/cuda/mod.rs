mod codegen;
pub mod cuda;
mod cuda_core;
pub use cuda::*;
