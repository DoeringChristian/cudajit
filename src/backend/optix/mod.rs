mod codegen;
pub mod optix;
pub mod optix_core;

pub use optix::*;

#[cfg(test)]
mod test;
