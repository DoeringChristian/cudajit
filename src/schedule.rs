use std::collections::HashMap;
use std::fmt::Debug;
use std::sync::Arc;

use slotmap::DefaultKey;
use smallvec::{smallvec, SmallVec};

use crate::backend::Backend;
use crate::ir::{self, Ir, Op, ParamType, Ref, VarId, VarType, IR};

#[derive(Clone, Copy, Debug, Hash, PartialEq, Eq)]
pub struct SVarId(pub usize);
impl std::fmt::Display for SVarId {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}

///
/// A Representation of a variable used in the ScheduleIr.
/// This only holds data that is needed to compile the Kernel.
///
/// Variables are densly stored in the ScheduleIr, simplifying the compilation.
///
#[derive(Debug, Default)]
pub struct ScheduleVar {
    pub op: Op,
    pub deps: SmallVec<[SVarId; 4]>,
    pub ty: VarType,
    pub param_ty: ParamType,
    pub reg: usize,
    pub param_offset: usize,
    pub literal: u64,
    pub size: usize,
}

impl ScheduleVar {
    pub fn is_literal(&self) -> bool {
        self.op == Op::Literal
    }
}

///
/// Helper struct for printing register names.
/// <prefix><register_index>
///
pub struct Reg<'a>(pub &'a ScheduleVar);
impl<'a> std::fmt::Display for Reg<'a> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}{}", self.0.ty.prefix(), self.0.reg)
    }
}
impl ScheduleVar {
    ///
    /// Returns a helper struct (Reg) that can be displayed with the correct prefix for the
    /// variable type.
    ///
    pub fn reg(&self) -> Reg {
        Reg(self)
    }
    ///
    /// Returns the raw index of the register for this variable.
    ///
    pub fn reg_idx(&self) -> usize {
        self.reg
    }
}

///
/// Intermediate representation for scheduled variables
///
// #[derive(Default)]
pub struct ScheduleIr {
    vars: Vec<ScheduleVar>,
    params: Vec<u64>,
    n_regs: usize,
    // The index into the hashmap consists of
    // the variable id and an optional index for
    // reindexing.
    visited: HashMap<VarId, SVarId>,
    backend: Arc<dyn Backend>,
}

impl Debug for ScheduleIr {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("ScheduleIr")
            .field("vars", &self.vars)
            .field("params", &self.params)
            .field("n_regs", &self.n_regs)
            .field("visited", &self.visited)
            // .field("backend", &self.backend)
            .finish()
    }
}

impl ScheduleIr {
    pub fn new(backend: &Arc<dyn Backend>, first_register: usize, size: usize) -> Self {
        Self {
            n_regs: first_register,
            params: vec![size as _],
            vars: Default::default(),
            visited: Default::default(),
            backend: backend.clone(),
        }
    }
    pub fn ids(&self) -> impl Iterator<Item = SVarId> {
        (0..self.vars.len()).map(|i| SVarId(i))
    }
    pub fn var(&self, id: SVarId) -> &ScheduleVar {
        &self.vars[id.0]
    }
    pub fn reg(&self, id: SVarId) -> Reg {
        self.var(id).reg()
    }
    // pub fn lit(&self, id: SVarId) -> Literal {
    //     self.var(id).lit()
    // }
    pub fn n_params(&self) -> usize {
        self.params.len()
    }
    pub fn n_regs(&self) -> usize {
        self.n_regs
    }
    fn next_reg(&mut self) -> usize {
        let reg = self.n_regs;
        self.n_regs += 1;
        reg
    }
    pub fn size(&self) -> usize {
        self.params[0] as usize
    }
    pub fn params(&self) -> &[u64] {
        &self.params
    }
    pub fn params_mut(&mut self) -> &mut [u64] {
        &mut self.params
    }
    fn push_var(&mut self, var: ScheduleVar) -> SVarId {
        let id = SVarId(self.vars.len());
        self.vars.push(var);
        id
    }
    fn push_param(&mut self, param: u64) -> usize {
        let idx = self.params.len();
        self.params.push(param);
        idx
    }
    pub fn collect_vars(&mut self, ids: &[VarId]) {
        for id in ids {
            self.collect(*id);
        }
    }
    ///
    /// Traverse computation graph and collect variables into Schedule.
    ///
    /// If a gather operation is encountered, that only depends on trivial operations we can
    /// reindex it using the parameter idx.
    ///
    pub fn collect(&mut self, id: VarId) -> SVarId {
        if self.visited.contains_key(&id) {
            return self.visited[&id];
        }

        let var = ir::var(&Ref::borrow(id));

        let mut sv = ScheduleVar {
            op: var.op,
            ty: var.ty.clone(),
            deps: smallvec![],
            reg: self.next_reg(),
            param_ty: var.param_ty,
            param_offset: 0,
            literal: var.literal,
            size: var.size,
        };

        // Collect dependencies
        if var.stop_traversal {
            sv.deps = smallvec![]
        } else {
            sv.deps = var
                .deps
                .iter()
                .map(|id| self.collect(*id))
                .collect::<SmallVec<[_; 4]>>();
        };
        match var.param_ty {
            ParamType::Input => {
                if var.is_literal() {
                    sv.param_offset = self.push_param(var.literal);

                    // Literal is pushed via params => we can set it to zero so that snapshot
                    // testing works.
                    sv.literal = 0;
                } else {
                    sv.param_offset = self.push_param(var.buffer.as_ref().unwrap().as_ptr());
                }
            }
            ParamType::Output => {
                sv.param_offset = self.push_param(var.buffer.as_ref().unwrap().as_ptr());
            }
            ParamType::None => {}
        }

        let id = self.push_var(sv);

        id
    }
}
