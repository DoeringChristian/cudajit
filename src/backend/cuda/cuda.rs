use cust::module::{ModuleJitOption, OptLevel};
use cust::prelude::{Context, DeviceBuffer, Module};
use cust::stream::{Stream, StreamFlags};
use cust::util::SliceExt;

use crate::backend;
use crate::schedule::{SVarId, ScheduleIr};
use crate::var::*;
use std::fmt::{Debug, Write};
use std::sync::Arc;

#[derive(Debug)]
pub struct Backend {
    ctx: Arc<Context>,
    stream: Option<cust::stream::Stream>,
}
impl Backend {
    pub fn new() -> Self {
        Self {
            ctx: Arc::new(cust::quick_init().unwrap()),
            stream: Some(Stream::new(StreamFlags::NON_BLOCKING, None).unwrap()),
        }
    }
}

impl backend::Backend for Backend {
    fn new_kernel(&self) -> Box<dyn backend::Kernel> {
        Box::new(CUDAKernel {
            ctx: self.ctx.clone(),
            asm: Default::default(),
            module: Default::default(),
        })
    }
    fn first_register(&self) -> usize {
        CUDAKernel::FIRST_REGISTER
    }

    fn buffer_uninit(&self, size: usize) -> Arc<dyn backend::Buffer> {
        unsafe { cust::sys::cuCtxSetCurrent(self.ctx.as_raw()) };
        Arc::new(Buffer {
            buffer: vec![0u8; size].as_slice().as_dbuf().unwrap(),
            ctx: self.ctx.clone(),
        })
    }

    fn buffer_from_slice(&self, slice: &[u8]) -> Arc<dyn backend::Buffer> {
        unsafe { cust::sys::cuCtxSetCurrent(self.ctx.as_raw()) };
        Arc::new(Buffer {
            buffer: slice.as_dbuf().unwrap(),
            ctx: self.ctx.clone(),
        })
    }
    fn synchronize(&self) {
        unsafe { cust::sys::cuCtxSetCurrent(self.ctx.as_raw()) };
        self.stream.as_ref().unwrap().synchronize().unwrap();
    }
}

impl Drop for Backend {
    fn drop(&mut self) {
        self.stream.take();
    }
}

#[derive(Debug)]
pub struct Buffer {
    buffer: DeviceBuffer<u8>,
    ctx: Arc<Context>,
}
impl backend::Buffer for Buffer {
    fn as_ptr(&self) -> u64 {
        self.buffer.as_device_ptr().as_raw()
    }
    fn as_vec(&self) -> Vec<u8> {
        self.buffer.as_host_vec().unwrap()
    }
}

pub struct Texture {
    ctx: Arc<Context>,
}

pub struct CUDAKernel {
    ctx: Arc<Context>,
    pub asm: String,
    pub module: Option<Module>,
}

impl Debug for CUDAKernel {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("CUDAKernel")
            .field("asm", &self.asm)
            .finish()
    }
}

impl backend::Kernel for CUDAKernel {
    fn execute_async(&mut self, ir: &mut ScheduleIr) {
        unsafe { cust::sys::cuCtxSetCurrent(self.ctx.as_raw()) };

        let func = self
            .module
            .as_ref()
            .expect("Need to compile Kernel before we can execute it!")
            .get_function(Self::ENTRY_POINT)
            .unwrap();

        let (_, block_size) = func.suggested_launch_configuration(0, 0.into()).unwrap();

        let grid_size = (ir.size() as u32 + block_size - 1) / block_size;

        let stream =
            cust::stream::Stream::new(cust::stream::StreamFlags::NON_BLOCKING, None).unwrap();

        let mut params = vec![ir.size() as u64];
        params.extend(ir.buffers().iter().map(|b| b.as_ptr()));

        unsafe {
            stream
                .launch(
                    &func,
                    grid_size,
                    block_size,
                    0,
                    &[params.as_mut_ptr() as *mut std::ffi::c_void],
                )
                .unwrap();
        }

        // stream.synchronize().unwrap();
    }
    #[allow(unused_must_use)]
    fn assemble(&mut self, ir: &ScheduleIr) {
        self.asm.clear();
        let n_params = 1 + ir.buffers().len(); // Add 1 for size
        let n_regs = ir.n_regs();

        /* Special registers:
             %r0   :  Index
             %r1   :  Step
             %r2   :  Size
             %p0   :  Stopping predicate
             %rd0  :  Temporary for parameter pointers
             %rd1  :  Pointer to parameter table in global memory if too big
             %b3, %w3, %r3, %rd3, %f3, %d3, %p3: reserved for use in compound
             statements that must write a temporary result to a register.
        */

        writeln!(self.asm, ".version {}.{}", 8, 0);
        writeln!(self.asm, ".target {}", "sm_86");
        writeln!(self.asm, ".address_size 64");

        writeln!(self.asm, "");

        writeln!(self.asm, ".entry {}(", Self::ENTRY_POINT);
        writeln!(
            self.asm,
            "\t.param .align 8 .b8 params[{}]) {{",
            ((n_params + 1) * std::mem::size_of::<u64>())
        );
        writeln!(self.asm, "");

        writeln!(
            self.asm,
            "\t.reg.b8   %b <{n_regs}>; .reg.b16 %w<{n_regs}>; .reg.b32 %r<{n_regs}>;"
        );
        writeln!(
            self.asm,
            "\t.reg.b64  %rd<{n_regs}>; .reg.f32 %f<{n_regs}>; .reg.f64 %d<{n_regs}>;"
        );
        writeln!(self.asm, "\t.reg.pred %p <{n_regs}>;");
        writeln!(self.asm, "");

        write!(
            self.asm,
            "\tmov.u32 %r0, %ctaid.x;\n\
             \tmov.u32 %r1, %ntid.x;\n\
             \tmov.u32 %r2, %tid.x;\n\
             \tmad.lo.u32 %r0, %r0, %r1, %r2; // r0 <- Index\n"
        );

        writeln!(self.asm, "");

        writeln!(
            self.asm,
            "\t// Index Conditional (jump to done if Index >= Size)."
        );
        writeln!(
            self.asm,
            "\tld.param.u32 %r2, [params]; // r2 <- params[0] (Size)"
        );

        write!(
            self.asm,
            "\tsetp.ge.u32 %p0, %r0, %r2; // p0 <- r0 >= r2\n\
            \t@%p0 bra done; // if p0 => done\n\
            \t\n\
            \tmov.u32 %r3, %nctaid.x; // r3 <- nctaid.x\n\
            \tmul.lo.u32 %r1, %r3, %r1; // r1 <- r3 * r1\n\
            \t\n"
        );

        write!(self.asm, "body: // sm_{}\n", 86); // TODO: compute capability from device

        for id in ir.ids() {
            let var = ir.var(id);
            match var.param_ty {
                ParamType::None => self.assemble_var(ir, id),
                ParamType::Input => {
                    let param_offset = (var.param.unwrap() + 1) * 8;
                    // Load from params
                    writeln!(self.asm, "");
                    writeln!(self.asm, "\t// [{}]: {:?} =>", id, var);
                    if var.is_literal() {
                        writeln!(
                            self.asm,
                            "\tld.param.{} {}, [params+{}];",
                            var.ty.name_cuda(),
                            var.reg(),
                            param_offset
                        );
                        continue;
                    } else {
                        writeln!(self.asm, "\tld.param.u64 %rd0, [params+{}];", param_offset);
                    }
                    if var.size > 1 {
                        writeln!(
                            self.asm,
                            "\tmad.wide.u32 %rd0, %r0, {}, %rd0;",
                            var.ty.size()
                        );
                    }

                    if var.ty == VarType::Bool {
                        writeln!(self.asm, "\tld.global.cs.u8 %w0, [%rd0];");
                        writeln!(self.asm, "\tsetp.ne.u16 {}, %w0, 0;", var.reg());
                    } else {
                        writeln!(
                            self.asm,
                            "\tld.global.cs.{} {}, [%rd0];",
                            var.ty.name_cuda(),
                            var.reg(),
                        );
                    }
                }
                ParamType::Output => {
                    let param_offst = (var.param.unwrap() + 1) * 8;
                    self.assemble_var(ir, id);
                    // let offset = param_idx * 8;
                    write!(
                        self.asm,
                        "\n\t// Store:\n\
                        \tld.param.u64 %rd0, [params + {}]; // rd0 <- params[offset]\n\
                            \tmad.wide.u32 %rd0, %r0, {}, %rd0; // rd0 <- Index * ty.size() + \
                            params[offset]\n",
                        param_offst,
                        var.ty.size(),
                    );

                    if var.ty == VarType::Bool {
                        writeln!(self.asm, "\tselp.u16 %w0, 1, 0, {};", var.reg());
                        writeln!(self.asm, "\tst.global.cs.u8 [%rd0], %w0;");
                    } else {
                        writeln!(
                                self.asm,
                                "\tst.global.cs.{} [%rd0], {}; // (Index * ty.size() + params[offset])[Index] <- var",
                                var.ty.name_cuda(),
                                var.reg(),
                            );
                    }
                }
            }
        }

        // End of kernel:

        writeln!(self.asm, "\n\t//End of Kernel:");
        writeln!(
            self.asm,
            "\n\tadd.u32 %r0, %r0, %r1; // r0 <- r0 + r1\n\
            \tsetp.ge.u32 %p0, %r0, %r2; // p0 <- r1 >= r2\n\
            \t@!%p0 bra body; // if p0 => body\n\
            \n"
        );
        writeln!(self.asm, "done:");
        write!(
            self.asm,
            "\n\tret;\n\
        }}\n"
        );

        std::fs::write("/tmp/tmp.ptx", &self.asm).unwrap();

        log::trace!("{}", self.asm);
    }

    fn compile(&mut self) {
        self.module = Some(
            Module::from_ptx(
                &self.asm,
                &[
                    ModuleJitOption::OptLevel(OptLevel::O0),
                    ModuleJitOption::GenenerateDebugInfo(true),
                    ModuleJitOption::GenerateLineInfo(true),
                ],
            )
            .unwrap(),
        );
    }

    fn assembly(&self) -> &str {
        &self.asm
    }
}

impl CUDAKernel {
    const ENTRY_POINT: &str = "cujit";
    const FIRST_REGISTER: usize = 4;
    #[allow(unused_must_use)]
    fn assemble_var(&mut self, ir: &ScheduleIr, id: SVarId) {
        super::codegen::assemble_var(&mut self.asm, ir, id);
    }
}

#[cfg(test)]
mod test {
    use crate::jit::Jit;
    use crate::trace::{ReduceOp, Trace};
    use crate::{jit, trace};

    #[test]
    fn refcounting() {
        let ir = Trace::default();
        ir.set_backend("cuda");

        let x = ir.buffer_f32(&[1.; 10]);
        assert_eq!(x.var().rc, 1, "rc of x should be 1 (in x)");
        let y = x.add(&x);
        // let y = ir::add(&x, &x);

        assert_eq!(
            x.var().rc,
            3,
            "rc of x should be 3 (2 refs in y and 1 in x)"
        );

        assert_eq!(y.var().rc, 1, "rc of y should be 1 (in y)");

        ir.schedule(&[&y]);
        let mut jit = Jit::default();

        assert_eq!(
            y.var().rc,
            2,
            "rc of y should be 2 (1 in y and 1 in schedule)"
        );

        jit.eval(&mut ir.borrow_mut());

        assert_eq!(
            x.var().rc,
            1,
            "rc of x should be 1 (dependencies of y shuld be cleaned)"
        );
        assert_eq!(
            y.var().rc,
            1,
            "rc of y should be 2 (y from schedule should be cleaned)"
        );

        insta::assert_snapshot!(jit.kernel_debug());

        assert_eq!(y.to_host_f32(), vec![2f32; 10]);
    }
    #[test]
    fn load_add_f32() {
        let ir = Trace::default();
        ir.set_backend("cuda");

        let x = ir.buffer_f32(&[1.; 10]);
        // let y = ir::add(&x, &x);
        let y = x.add(&x);

        ir.schedule(&[&y]);
        let mut jit = Jit::default();
        jit.eval(&mut ir.borrow_mut());

        insta::assert_snapshot!(jit.kernel_debug());

        assert_eq!(y.to_host_f32(), vec![2f32; 10]);
    }
    #[test]
    fn load_gather_f32() {
        let ir = Trace::default();
        ir.set_backend("cuda");

        let x = ir.buffer_f32(&[1., 2., 3., 4., 5.]);
        let i = ir.buffer_u32(&[0, 1, 4]);
        let y = x.gather(&i, None);

        ir.schedule(&[&y]);
        let mut jit = Jit::default();
        jit.eval(&mut ir.borrow_mut());

        insta::assert_snapshot!(jit.kernel_debug());

        assert_eq!(y.to_host_f32(), vec![1., 2., 5.]);
    }
    #[test]
    fn reindex() {
        let ir = Trace::default();
        ir.set_backend("cuda");

        let x = ir.index(10);

        let i = ir.index(3);
        let c = ir.literal_u32(2);
        let i = i.add(&c);

        let y = x.gather(&i, None);

        ir.schedule(&[&y]);
        let mut jit = Jit::default();
        jit.eval(&mut ir.borrow_mut());

        insta::assert_snapshot!(jit.kernel_debug());

        assert_eq!(y.to_host_u32(), vec![2, 3, 4]);
    }
    #[test]
    fn index() {
        let ir = Trace::default();
        ir.set_backend("cuda");

        let i = ir.index(10);

        ir.schedule(&[&i]);
        let mut jit = Jit::default();
        jit.eval(&mut ir.borrow_mut());

        insta::assert_snapshot!(jit.kernel_debug());

        assert_eq!(i.to_host_u32(), vec![0, 1, 2, 3, 4, 5, 6, 7, 8, 9]);
    }
    #[test]
    fn gather_eval() {
        let ir = Trace::default();
        ir.set_backend("cuda");

        let tmp_x;
        let tmp_y;
        let tmp_z;
        let r = {
            let x = ir.index(3);
            dbg!();
            assert_eq!(x.var().rc, 1);
            let y = ir.buffer_u32(&[1, 2, 3]);
            dbg!();
            assert_eq!(y.var().rc, 1);

            // let z = ir::add(&x, &y);
            let z = x.add(&y);
            dbg!();
            assert_eq!(x.var().rc, 2);
            assert_eq!(y.var().rc, 2);
            assert_eq!(z.var().rc, 1);

            let r = z.gather(&ir.index(3), None);
            dbg!();
            assert_eq!(x.var().rc, 2);
            assert_eq!(y.var().rc, 2);
            assert_eq!(z.var().rc, 3);
            assert_eq!(r.var().rc, 1);
            tmp_x = x.id();
            tmp_y = y.id();
            tmp_z = z.id();
            r
        };
        assert_eq!(r.var().rc, 1);
        assert_eq!(ir.borrow_mut().get_var(tmp_x).unwrap().rc, 1);
        assert_eq!(ir.borrow_mut().get_var(tmp_y).unwrap().rc, 1);
        assert_eq!(ir.borrow_mut().get_var(tmp_z).unwrap().rc, 2); // z is referenced by r and the
                                                                   // schedule

        ir.schedule(&[&r]);
        let mut jit = Jit::default();
        jit.eval(&mut ir.borrow_mut());

        assert_eq!(r.var().rc, 1);
        assert!(ir.borrow_mut().get_var(tmp_z).is_none());

        insta::assert_snapshot!(jit.kernel_debug());

        assert_eq!(r.to_host_u32(), vec![1, 3, 5]);
    }
    #[test]
    fn paralell() {
        let ir = Trace::default();
        ir.set_backend("cuda");

        let x = ir.index(10);

        let y = ir.index(3);

        ir.schedule(&[&x, &y]);
        let mut jit = Jit::default();
        jit.eval(&mut ir.borrow_mut());

        insta::assert_snapshot!(jit.kernel_debug());

        assert_eq!(x.to_host_u32(), vec![0, 1, 2, 3, 4, 5, 6, 7, 8, 9]);
        assert_eq!(y.to_host_u32(), vec![0, 1, 2]);
    }
    #[test]
    fn load_gather() {
        let ir = Trace::default();
        ir.set_backend("cuda");

        let x = ir.buffer_f32(&[1., 2., 3.]);

        ir.schedule(&[&x]);

        let mut jit = Jit::default();
        jit.eval(&mut ir.borrow_mut());

        insta::assert_snapshot!(jit.kernel_debug());

        assert_eq!(x.to_host_f32(), vec![1., 2., 3.]);
    }
    #[test]
    fn eval_scatter() {
        let ir = Trace::default();
        ir.set_backend("cuda");

        let x = ir.buffer_u32(&[0, 0, 0, 0]);
        let c = ir.literal_u32(1);
        let x = x.add(&c); // x: [1, 1, 1, 1]

        let i = ir.index(3);
        let c = ir.literal_u32(1);
        let i = i.add(&c); // i: [1, 2, 3]

        let y = ir.literal_u32(2);

        y.scatter(&x, &i, None); // x: [1, 2, 2, 2]

        let mut jit = Jit::default();
        jit.eval(&mut ir.borrow_mut());

        insta::assert_snapshot!(jit.kernel_debug());

        assert_eq!(x.to_host_u32(), vec![1, 2, 2, 2]);
    }
    #[test]
    fn scatter_twice() {
        let ir = Trace::default();
        ir.set_backend("cuda");

        let x = ir.buffer_u32(&[0, 0, 0, 0]);

        let i = ir.index(3);
        let c = ir.literal_u32(1);
        let i = i.add(&c);

        let y = ir.literal_u32(2);

        y.scatter(&x, &i, None);

        let i = ir.index(2);

        let y = ir.literal_u32(3);

        y.scatter(&x, &i, None);

        let mut jit = Jit::default();
        jit.eval(&mut ir.borrow_mut());

        insta::assert_snapshot!(jit.kernel_debug());

        assert_eq!(x.to_host_u32(), vec![3, 3, 2, 2]);
    }
    #[test]
    fn scatter_twice_add() {
        let ir = Trace::default();
        ir.set_backend("cuda");

        let x = ir.buffer_u32(&[0, 0, 0, 0]);

        let i = ir.index(3);
        let c = ir.literal_u32(1);
        let i = i.add(&c);

        let y = ir.literal_u32(2);

        y.scatter(&x, &i, None);

        let i = ir.index(2);

        let y = ir.literal_u32(3);

        y.scatter(&x, &i, None);

        let c = ir.literal_u32(1);
        let x = x.add(&c);

        ir.schedule(&[&x]);

        let mut jit = Jit::default();
        jit.eval(&mut ir.borrow_mut());

        insta::assert_snapshot!(jit.kernel_debug());

        assert_eq!(x.to_host_u32(), vec![4, 4, 3, 3]);
    }
    #[test]
    fn scatter_reduce() {
        let ir = Trace::default();
        ir.set_backend("cuda");

        let x = ir.buffer_u32(&[0, 0, 0, 0]);

        let i = ir.buffer_u32(&[0, 0, 0]);

        let y = ir.literal_u32(1);

        y.scatter_reduce(&x, &i, None, ReduceOp::Add);

        ir.schedule(&[&x]);

        let mut jit = Jit::default();
        jit.eval(&mut ir.borrow_mut());

        insta::assert_snapshot!(jit.kernel_debug());

        assert_eq!(x.to_host_u32(), vec![3, 0, 0, 0]);
    }
}
