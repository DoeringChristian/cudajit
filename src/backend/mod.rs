pub mod cuda;
pub mod optix;

use std::any::Any;
use std::fmt::Debug;
use std::sync::Arc;

use crate::schedule::{Env, ScheduleIr};

pub trait Texture: Debug + Sync + Send {
    fn as_any(&self) -> &dyn Any;
    fn channels(&self) -> usize;
    fn dimensions(&self) -> usize;
    fn shape(&self) -> &[usize];
    fn copy_from_buffer(&self, buf: &dyn Buffer);
    fn copy_to_buffer(&self, buf: &dyn Buffer);
}

pub trait Kernel: Debug + Sync + Send {
    fn as_any(&self) -> &dyn Any;
    fn execute_async(&mut self, ir: &mut Env, size: usize) -> Arc<dyn DeviceFuture>;
    fn assembly(&self) -> &str;
}

pub trait DeviceFuture: Debug + Sync + Send {
    fn wait(&self);
}

pub trait Buffer: Debug + Sync + Send {
    fn as_any(&self) -> &dyn Any;
    fn copy_to_host(&self, dst: &mut [u8]);
    fn ptr(&self) -> Option<u64>;
}

pub trait Backend: Debug + Sync + Send {
    fn as_any(&self) -> &dyn Any;
    fn as_any_mut(&mut self) -> &mut dyn Any;
    fn compile_kernel(&self, ir: &ScheduleIr, env: &Env) -> Box<dyn Kernel>;
    fn create_texture(&self, shape: &[usize], n_channels: usize) -> Arc<dyn Texture>;
    fn buffer_uninit(&self, size: usize) -> Arc<dyn Buffer>;
    fn buffer_from_slice(&self, slice: &[u8]) -> Arc<dyn Buffer>;
    fn first_register(&self) -> usize;
    fn synchronize(&self);
    fn create_accel(&self, desc: AccelDesc) -> Arc<dyn Accel>;
    fn set_compile_options(&mut self, compile_options: &CompileOptions);
    fn set_miss_from_str(&mut self, entry_point: &str, source: &str);
    fn push_hit_from_str(&mut self, entry_point: &str, source: &str);
    fn ident(&self) -> &'static str;
}

pub trait Accel: Debug + Sync + Send {
    fn as_any(&self) -> &dyn Any;
}

#[derive(Clone, Default)]
pub struct CompileOptions {
    pub num_payload_values: i32,
}

pub enum GeometryDesc {
    Triangles {
        vertices: Arc<dyn Buffer>,
        indices: Arc<dyn Buffer>,
    },
}
pub struct InstanceDesc {
    pub geometry: usize,
    pub transform: [f32; 12],
}
pub struct AccelDesc<'a> {
    pub geometries: &'a [GeometryDesc],
    pub instances: &'a [InstanceDesc],
}
