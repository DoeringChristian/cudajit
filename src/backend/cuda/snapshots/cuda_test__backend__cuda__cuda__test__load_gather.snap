---
source: src/backend/cuda/cuda.rs
expression: jit.kernel_debug()
---

